composer/checkpoint/save.py

```diff
@@ -0,0 +1,133 @@
+    overwrite: bool = False,
+    save_format: str = 'pt',  # or hf, safetensor
+) -> Optional[str]:
+    """Saves a state dict to local disk.
+
+    Args:
+        state_dict (Dict[str,Any]): The state dict to save.
+        destination_dir (str): The directory to save the state dict to.
+        filename (str): The name of the file to save the state dict to.
+        overwrite (bool): If True, the file will be overwritten if it exists.
+        save_format (str): The format to save the state dict in. One of 'pt', 'hf', or 'safetensor'.
+        async_save (bool): If True, the save will be done asynchronously and the function will return with the path of where it was going to be saved
+
+    Returns:
+        str: The full path to the saved state dict if sharded is false and rank 0 or if sharded is true, otherwise None.
+    """
+    if state_dict == {}:
+        return None
+    sharded_state_dict = is_state_dict_sharded(state_dict)
+
+    if sharded_state_dict:
+        path_saved = _save_sharded_state_dict_to_disk(state_dict, destination_file_path, overwrite, save_format)
+    else:
+        if dist.get_global_rank() == 0:
+            path_saved = _save_full_state_dict_to_disk(state_dict, destination_file_path, overwrite, save_format)
+        else:
+            path_saved = None
+
+    return path_saved
+
+
+def _save_sharded_state_dict_to_disk(
+    state_dict: Dict[str, Any],
+    destination_file_path: str,
+    overwrite: bool = False,
+    save_format: str = 'pt',  # or safetensor
+) -> Optional[str]:
+
+    if save_format != 'pt':
+        raise NotImplementedError(f'Saving sharded state dict to disk in format {save_format} is not supported.')
+
+    if state_dict == {}:
+        return None
+
+    # If used specifies filename instead of directory suffixes, strip them and warn.
+    if len(Path(destination_file_path).suffixes) > 0:
+        stripped_path = _strip_suffixes(destination_file_path)
+        warnings.warn(
+            textwrap.dedent(
+                f"""Sharded checkpoints require a directory path not a file path:
+            {destination_file_path} will have its extensions stripped and checkpoints will be saved in {stripped_path}
+            as {stripped_path}/{_TORCH_DISTRIBUTED_CHECKPOINTS_FILENAME}""",
+            ),
+        )
+        destination_file_path = stripped_path
+
+    if dist.get_global_rank() == 0 and not overwrite and os.path.exists(destination_file_path):
+        raise ValueError(f'Directory {destination_file_path} already exists. Set overwrite=True to overwrite it.')
+
+    log.debug(
+        f'Starting saving of sharded state dict to {destination_file_path}/{_TORCH_DISTRIBUTED_CHECKPOINTS_FILENAME}',
+    )
+
+    # For 2.3.0 and above you can use checkpoint_id, but this version works the best for all versions
+    # of torch (and makes pyright happier) that we support, so we use it for now.
+    DCP.save(state_dict=state_dict, storage_writer=DCP.FileSystemWriter(destination_file_path))
```

Comment:
no way to supply planner? or process group?

Reply:
we can add that in a future PR. For now (with this tight timeline), we just want a simple mvp for these API's

Reply:
I can add them back in if you're ok with no unit tests for that functionality

Reply:
Follow on PR is fine, pro modularity. Note that this blocks switchover due to a few issues which require planner CC: @bigning  on adam

Reply:
ah, that's a good point. We definitely need to add [this](https://github.com/mosaicml/composer/blob/ba82cc9ec943e316da59ec6ec0ba540008c9ca1f/composer/utils/checkpoint.py#L1156) in follow on PR before enabling V2 in production.